watertap/unit_models/ion_exchange_0D.py

```diff
@@ -0,0 +1,1352 @@
+            prop_in = b.properties_in[0]
+            return (
+                b.sfr
+                == pyunits.convert(
+                    prop_in.flow_vol_phase["Liq"], to_units=pyunits.m**3 / pyunits.hr
+                )
+                / b.bed_vol
+            )
+
+        @self.Constraint(doc="Bed area calculation")
+        def eq_bed_area(b):
+            return b.bed_area == b.pi * (b.bed_diam / 2) ** 2
+
+        @self.Constraint(doc="Bed volume calculation")
+        def eq_bed_vol(b):
+            return self.bed_vol == self.bed_area * self.bed_depth
+
+        @self.Constraint(doc="Bed depth to bed diameter ratio")
+        def eq_bed_depth_to_diam_ratio(b):
+            return b.bed_depth / b.bed_diam >= b.bed_depth_to_diam_ratio
+
+        @self.Constraint(doc="Column height")
+        def eq_col_height(b):
+            return (
+                b.col_height
+                == b.bed_depth + b.distributor_h + b.underdrain_h + b.bed_expansion_h
+            )
+
+        @self.Constraint(doc="Column vol")
+        def eq_col_vol(b):
+            return b.col_vol == b.col_height * b.bed_area
+
+        # =========== KINETICS ===========
+        @self.Constraint(ion_set, doc="Fluid mass transfer coeff")
+        def eq_fluid_mass_transfer_coeff(b, j):
+            prop_in = b.properties_in[0]
+            return (
+                b.fluid_mass_transfer_coeff[j]
+                == (prop_in.diffus_phase_comp["Liq", j] * b.Sh[j]) / b.resin_diam
+            )
+
+        @self.Constraint(ion_set, doc="Rate coefficient")
+        def eq_rate_coeff(b, j):
+            return b.rate_coeff[j] == (
+                6 * (1 - b.bed_porosity) * b.fluid_mass_transfer_coeff[j]
+            ) / (
+                pyunits.convert(b.resin_bulk_dens, to_units=pyunits.kg / pyunits.m**3)
+                * b.resin_diam
+            )
+
+        @self.Constraint(ion_set, doc="Height of transfer unit - HTU")
+        def eq_HTU(b, j):
+            return b.HTU[j] == b.vel_bed / (
+                pyunits.convert(b.resin_bulk_dens, to_units=pyunits.kg / pyunits.m**3)
+                * b.rate_coeff[j]
+            )
+
+        @self.Constraint(doc="Partition ratio")
+        def eq_partition_ratio(b):
+            prop_in = b.properties_in[0]
+            return b.partition_ratio == (
+                b.resin_eq_capacity * b.resin_bulk_dens
+            ) / pyunits.convert(
+                sum(prop_in.conc_equiv_phase_comp["Liq", j] for j in ion_set),
+                to_units=pyunits.mol / pyunits.L,
+            )
+
+        @self.Constraint(doc="Left hand side of constant pattern sol'n")
+        def eq_lh(b):
+            return b.lh == b.num_transfer_units * (b.dimensionless_time - 1)
+
+        @self.Constraint(ion_set, doc="Right hand side of fixed pattern soln")
+        def eq_fixed_pattern_soln(b, j):
+            return (
+                b.lh
+                == (log(b.c_norm[j]) - b.R_eq[j] * log(1 - b.c_norm[j]))
```

Comment:
Here you are using the `pyomo.environ log` function. This is going to have some pretty poor convergence behavior if your `b.c_norm` vars are very small. You may want to consider using a so-called "safe log" function. 

I believe IDAES has a "safe-log" function. @andrewlee94 may have some input here.